File: 1-two-sum.py
# Two sum (https://leetcode.com/problems/two-sum/)


class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # 1. Create an empty dict
        num_dict = dict()
        for index in range(len(nums)):
            # 3. While iterating subtract current num from target
            # and check if dict has the needed num (result of substraction)
            needed_num = target - nums[index]
            if needed_num in num_dict:
                # 4. If it has, return the current index and index of needed num
                return [index, num_dict[needed_num]]
            # 2. Iterate over nums and fill the dict with pairs "num": "index of num"
            num_dict[nums[index]] = index

File: 1-two-sum.ts
// Two Sum (https://leetcode.com/problems/two-sum/)

function twoSum(nums: number[], target: number): number[] {
    let result: number[] = []
    // 1. Create a map
    const numMap = new Map()
    nums.forEach((number, index) => {
        // 3. While iterating subtract current num from target
        // and check if map has the needed num (result of substraction)
        const neededNumber = target - number
        if (numMap.has(neededNumber)) {
            // 4. Return the current index and index of needed num
            return (result = [index, numMap.get(neededNumber)])
        }
        // 2. Iterate over nums and fill the map with pairs "num": "index of num"
        numMap.set(number, index)
    })
    return result
}
